vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,917 @@
+  metrics.int_metrics_list.emplace_back("current_champ", current_champ);
+  for (size_t stride = 0; stride < scores.size(); ++stride)
+  { scores[stride].persist(metrics, "_" + std::to_string(stride)); }
+}
+
+// This sets up example with correct ineractions vector
+void interaction_config_manager::apply_config(example* ec, size_t stride)
+{
+  if (ec == nullptr) { return; }
+  if (stride < max_live_configs) { ec->interactions = &(scores[stride].live_interactions); }
+  else
+  {
+    THROW("fatal (automl): trying to apply a config higher than max configs allowed");
+  }
+}
+
+void interaction_config_manager::revert_config(example* ec) { ec->interactions = nullptr; }
+
+template <typename CMType, bool is_explore>
+void predict_automl(automl<CMType>& data, multi_learner& base, multi_ex& ec)
+{
+  size_t champ_stride = data.cm->current_champ;
+  for (example* ex : ec) { data.cm->apply_config(ex, champ_stride); }
+
+  auto restore_guard = VW::scope_exit([&data, &ec, &champ_stride] {
+    for (example* ex : ec) { data.cm->revert_config(ex); }
+  });
+
+  base.predict(ec, champ_stride);
+}
+
+// inner loop of learn driven by # MAX_CONFIGS
+template <typename CMType>
+void offset_learn(automl<CMType>& data, multi_learner& base, multi_ex& ec, const size_t stride, CB::cb_class& logged,
+    size_t labelled_action)
+{
+  for (example* ex : ec) { data.cm->apply_config(ex, stride); }
+
+  auto restore_guard = VW::scope_exit([&data, &ec, &stride] {
+    for (example* ex : ec) { data.cm->revert_config(ex); }
+  });
+
+  if (!base.learn_returns_prediction) { base.predict(ec, stride); }
+
+  base.learn(ec, stride);
+
+  const auto& action_scores = ec[0]->pred.a_s;
+  // cb_adf => first action is a greedy action
+  const auto maxit = action_scores.begin();
+  const uint32_t chosen_action = maxit->action;
+
+  // extra asserts
+  assert(chosen_action < ec.size());
+  assert(labelled_action < ec.size());
+
+  const float w = logged.probability > 0 ? 1 / logged.probability : 0;
+  const float r = -logged.cost;
+
+  data.cm->scores[stride].update(chosen_action == labelled_action ? w : 0, r);
+
+  // cache the champ
+  if (data.cm->current_champ == stride) { data.champ_a_s = std::move(ec[0]->pred.a_s); }
+}
+
+// this is the registered learn function for this reduction
+// mostly uses config_manager and actual_learn(..)
+template <typename CMType, bool is_explore>
+void learn_automl(automl<CMType>& data, multi_learner& base, multi_ex& ec)
+{
+  assert(data.all->weights.sparse == false);
+
+  bool is_learn = true;
+
+  if (is_learn) { data.cm->total_learn_count++; }
+  // extra assert just bc
+  assert(data.all->interactions.empty() == true);
+
+  CB::cb_class logged{};
+  size_t labelled_action = 0;
+  if (is_learn)
+  {
+    const auto it = std::find_if(ec.begin(), ec.end(), [](example* item) { return !item->l.cb.costs.empty(); });
+
+    if (it != ec.end())
+    {
+      logged = (*it)->l.cb.costs[0];
```

Comment:
is this correct? if `cb.costs.empty()` is true that means that there maybe is a label but it could just be the shared label here, no?